diff-containers/README.md

```diff
@@ -1 +1,238 @@
+  {- omitted for now -}
+```
+
+The way in which subtraction is facilitated is by implementing the `invert`
+method in a `Group` instance. As a consequence, not only do we have to add
+inverses for diffs, diff histories and diff entries, we must also define how
+these inverses are handled by summing.
+
+First, we start by adding inverse elements to the definition of `DiffEntry`. One
+may think `Insert` and `Delete` are already inverses of one another, and that
+would be correct in terms of adhering to the type class laws, but it would have
+adverse effects on the semantics of applying diffs to maps. Consider what would
+happen if we were to consider `Insert` and `Delete` inverses in the following
+example. Ignore diff histories for now:
+
+```haskell
+d1 = fromList [(0, Delete 17)]
+d2 = fromList [(0, Insert 42)]
+d3 = d1 <> d2 = fromList []
+```
+
+Applying `d3` to a map would behave like the identity function. Applying `d1`
+and `d2` consecutively would not behave like the identity function, as it would
+map `0` to `42`. Through some abuse of the definition of distributivity, we can
+say that applying diffs does not distribute over the summing of diffs.
+
+Instead, we add explicit inverse constructors for both `Insert` and `Delete`. We
+do not make `Semigroup`/`Monoid`/`Group` instances for diff entries, however. We
+shall see that we only need these 4 constructors in order to provide the
+instances for diff histories.
+
+```haskell
+data DiffEntry v =
+    Insert !v
+  | Delete !v
+  | UnsafeAntiInsert !v -- ^ Unsafe!
+  | UnsafeAntiDelete !v -- ^ Unsafe!
+```
+
+Inverse diff entries will *cancel out* when we sum diff histories. For example:
+
+```haskell
+h1 = fromList [Insert 1, UnsafeAntiDelete 2, UnsafeAntiInsert 3]
+h2 = fromList [Delete 2, Insert 3, Insert 3]
+h3 = h1 <> h2 = [Insert 1, Delete 2, Insert 3]
+```
+
+How do we invert diff histories? The type class laws that a `Group` instance
+should adhere to give us a clue:
+
+```haskell
+-- left inverse
+a <> invert a == mempty
+-- right inverse
+invert a <> a == mempty
+```
+
+Given that inverse diff entries cancel each other out, inverting a diff history should invert all diff entries and reverse their order. See this example:
+
+```haskell
+h        = fromList [Insert 1, Delete 2, Delete 3, Insert 4]
+invert h = fromList [UnsafeAntiInsert 4, UnsafeAntiDelete 3, UnsafeAntiDelete 2, UnsafeAntiInsert 1]
+
+invert h <> h == mempty
+h <> invert h == mempty
+```
+
+Summing of diffs follows from pairwise summing the inner diff histories.
+Inverting diffs follows from inverting each inner diff history.
+
+### 4) Applying diffs with inverses: positivity and normality
+
+It is an unfortunate side effect of facilitating a `Group` instance for diffs
+that we can get into situations where applying diffs will fail or produce wrong
+results due to diffs containing internally unresolved sums and subtractions. The
+responsibility of downstream code is to ensure that diffs that are applied are
```

Comment:
Why is this a responsibility of the downstream code? Can't we maintain these invariants?

Reply:
Unfortunately, I don't think we can: applying any diff that has any `UnsafeAnti*` constructors can go wrong, and there's no way from keeping downstream code from misusing subtraction/inversion to create unsafe diffs that contain these constructors. Unless, maybe, by running expensive and non-trivial run-time checks.

I could add assertions to the code that do check these invariants in tests, but they should probably not be enabled in "production mode". I'll add a separate issue for this.

Reply:
See https://github.com/input-output-hk/anti-diffs/issues/26